Start framework for kl_divergence(-,-) in torch.distributions

fritzo opened the pull request:
This adds a generic function `kl_divergence(-,-)` and a registration decorator
```py
@register_kl(Normal, Normal)
def kl_normal_normal(p, q):
    ...
```
such that `kl_divergence(Normal(0, 1), Normal(0, 2))` will call `_kl_normal_normal()` under the hood.

The implementation has amortized constant lookup time, including failed lookup for `NotImplementedError`. It is important to keep this cheap so that generic inference algorithms can
```py
try:
    kl = kl_divergence(p, q):
except NotImplementedError:
    ...use an approximate method...
```

This also adds a generic `TestKL` to which test cases can be added.

Addresses #4330, probtorch#37

fritzo committed:
Add kl_divergence(-,-) registration mechanism

fritzo committed:
Memoize kl lookup

fritzo committed:
Also memoize NotImplementedErrors

fritzo committed:
Add kl_divergence docs

fritzo committed:
Add _kl_gamma_gamma and a generic test

fritzo committed:
Improve documentation

fritzo commented:
cc @apaszke @vishwakftw

fritzo committed:
Simplify register_kl() logic

fritzo commented on torch/distributions/gamma.py:
```diff
@@ -54,3 +55,12 @@ def log_prob(self, value):
+
+
+@register_kl(Gamma, Gamma)
+def _kl_gamma_gamma(p, q):
```
This is only intended to demonstrate usage. @vishwakftw may implement a more efficient version of this.

fritzo replied:
Maybe this should be public so users aren't force to use PyTorch's `kl_divergence()` registration mechanism?

fritzo committed:
Make kl_gamma_gamma public; fix docs

fritzo committed:
Make test_kl_monte_carlo work with batched parameters

apaszke commented on torch/distributions/kl.py:
```diff
@@ -0,0 +1,83 @@
+    except KeyError:
+        pass
+    # Look for the first approximate match.
+    for super_p, super_q in _KL_REGISTRY:
```
I guess that's ok as long as we keep the class hierarchy flat (depth <= 2), but we should do something better if we ever plan to have longer inheritance chains

fritzo replied:
Sure, let's see what happens when things get more complex. The current strategy works for our depth=3 hierarchy `Distribution` -> `Gamma` -> `Chi2`. I seems safe since in the worst case a user can explicitly register all desired pairs.

fritzo replied:
I'll try to switch to most-specific match assuming single inheritance...

fritzo replied:
Done.

apaszke commented on torch/distributions/gamma.py:
```diff
@@ -54,3 +55,12 @@ def log_prob(self, value):
+
+
+@register_kl(Gamma, Gamma)
+def kl_gamma_gamma(p, q):
```
Why should this be public?

fritzo replied:
The main reason is to allow users to use these implementations for user defined distributions. E.g.
```py
class LogGamma(Distribution):
    def __init__(self, alpha, beta):
        self._gamma = Gamma(alpha, beta)
    def rsample(self):
        return self._gamma.rsample().log()

@register_kl
def kl_loggamma_loggamma(p, q):
    return kl_gamma_gamma(p._gamma, q._gamma)
```
If the `_kl_gamma_gamma()` were private, then `kl_loggamma_loggamma` would need to use `kl_divergence` under the hood, and we'd be hitting the registry twice each call.

apaszke replied:
Hmm I still think it would just be nicer to have one consistent API instead of a bunch of methods scattered over files. How about this: you can either use `distributions.kl`, or we add an option to lookup functions for types, and you can cache the function wherever you want. Something like:

```python
kl_gamma_gamma = distributions.kl(Gamma, Gamma)
@register_kl(LogGamma, LogGamma)
def kl_loggamma_loggamma(p, q):
    return kl_gamma_gamma(p._gamma, q._gamma)
```

vishwakftw replied:
I think that is what @fritzo discussed in the distributions channel recently. I also think it would be better to have all of them in one place.

fritzo replied:
@apaszke Good point, I've added a `kl_function` lookup method as you suggest.

fritzo committed:
Find the most specific match ordered by subtype

fritzo committed:
Avoid warning about ambuguity when functions are the same

vishwakftw commented on torch/distributions/__init__.py:
```diff
@@ -34,11 +34,12 @@
 from .distribution import Distribution
 from .exponential import Exponential
 from .gamma import Gamma
-from .chi2 import Chi2
```
Will not adding `.chi2` before `.gamma` cause circular import as @alicanb commented on my earlier PR where I did something similar?

fritzo replied:
No this is fine. The Gamma module we be loaded (but not imported) at the line `from .chi2 import Chi2`. The import order in this PR was automatically determined by `isort` (which saves a ton of time BTW).

apaszke commented on torch/distributions/kl.py:
```diff
@@ -0,0 +1,114 @@
+        return NotImplemented
+    # Check that the left- and right- lexicographic orders agree.
+    left_p, left_q = min(_Match(*m) for m in matches).types
+    right_q, right_p = min(_Match(*reversed(m)) for m in matches).types
```
That looks great!

fritzo replied:
Thanks for pushing for a better implementation :wink:

apaszke reviewed (approved):
LGTM. Let me know what you want to do with specific overloads

fritzo committed:
Expose kl_function(type_p,type_q) for function lookup

fritzo committed:
Merge branch 'master' into register-kl-divergence

fritzo committed:
Remove kl_function() and hide _kl_gamma_gamma

fritzo commented:
Ok I've removed `kl_function()` and moved `_kl_gamma_gamma()` into `kl.py`. @vishwakftw apologies for merge conflicts, but this PR should now be closer to your follow-up PR.

vishwakftw commented:
@fritzo No worries. I think as soon as this is merged into master, I can issue mine. Thank you.

fritzo committed:
Merge branch 'master' into register-kl-divergence
